File: asteroids_libgdx/core/src/dev/ian/assroids/screen/BaseScreen.java
package dev.ian.assroids.screen;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Screen;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;

import dev.ian.assroids.AssRoids;
import dev.ian.assroids.asset.Asset;

/**
 * Created by: Ian Parcon
 * Date created: Sep 15, 2018
 * Time created: 9:10 AM
 */
public abstract class BaseScreen implements Screen {

    protected SpriteBatch batch;
    protected AssRoids game;
    protected Asset asset;
    protected float width;
    protected float height;

    public BaseScreen(AssRoids game) {
        this.game = game;
        batch = game.getBatch();
        asset = Asset.instance();
        width = Gdx.graphics.getWidth();
        height = Gdx.graphics.getHeight();
    }

    public abstract void update(float delta);

    @Override
    public void render(float delta) {
        update(delta);
        Gdx.gl.glClearColor(40f / 255f, 40f/ 255f, 40f/255f, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
    }

    @Override
    public void resize(int width, int height) {
        this.width = width;
        this.height = height;
    }

    @Override
    public void pause() {

    }

    @Override
    public void resume() {

    }

    @Override
    public void hide() {

    }

    @Override
    public void dispose() {
        asset.dispose();
        game.dispose();
    }
}

File: libgx_shark_adventure/core/src/com/nopalsoft/sharkadventure/screens/BaseScreen.kt
package com.nopalsoft.sharkadventure.screens

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.InputAdapter
import com.badlogic.gdx.InputMultiplexer
import com.badlogic.gdx.Screen
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.nopalsoft.sharkadventure.MainShark
import com.nopalsoft.sharkadventure.Settings.save

/**
 * The base logic shared by all screens in a game.
 */
abstract class BaseScreen(var game: MainShark) : InputAdapter(), Screen {
    var camera: OrthographicCamera
    var batcher: SpriteBatch
    var stage = game.stage

    init {
        stage.clear()
        batcher = game.batcher

        camera = OrthographicCamera(SCREEN_WIDTH, SCREEN_HEIGHT)
        camera.position.set(SCREEN_WIDTH / 2F, SCREEN_HEIGHT / 2F, 0F)

        val input = InputMultiplexer(stage, this)
        Gdx.input.inputProcessor = input
    }

    override fun render(delta: Float) {
        update(delta)
        stage.act(delta)

        camera.update()
        batcher.projectionMatrix = camera.combined
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)

        draw(delta)
        stage.draw()
    }

    abstract fun update(delta: Float)

    abstract fun draw(delta: Float)

    override fun resize(width: Int, height: Int) {
        stage.viewport.update(width, height, true)
    }

    override fun show() {
    }

    override fun hide() {
        save()
    }

    override fun pause() {
        save()
    }

    override fun resume() {
    }

    override fun dispose() {
    }

    companion object {
        const val SCREEN_WIDTH = 800F
        const val SCREEN_HEIGHT = 480F
        const val WORLD_WIDTH = 8F
        const val WORLD_HEIGHT = 4.8F
    }
}
